src/coreclr/jit/emit.cpp

```diff
@@ -9119,6 +9124,65 @@ void emitter::emitNxtIG(bool extend)
 #endif
 }
 
+//------------------------------------------------------------------------
+// emitRemoveLastInstruction: Remove the last instruction emitted; it has been optimized away by the
+// next instruction we are generating. `emitLastIns` must be non-null, meaning there is a
+// previous instruction. The previous instruction might have already been saved, or it might
+// be in the currently accumulating insGroup buffer.
+//
+// The `emitLastIns` is set to nullptr after this function. It is expected that a new instruction
+// will be immediately generated after this, which will set it again.
+//
+// NOTE: It is expected that the GC effect of the removed instruction will be handled by the newly
+// generated replacement(s).
+//
+void emitter::emitRemoveLastInstruction()
+{
+    assert(emitLastIns != nullptr);
+    assert(emitLastInsIG != nullptr);
+
+    JITDUMP("Removing saved instruction in %s:\n", emitLabelString(emitLastInsIG));
+    JITDUMPEXEC(dispIns(emitLastIns))
+
+    // We should assert it's not a jmp, as that would require updating the jump lists, e.g. emitCurIGjmpList.
+
+    BYTE* lastInsActualStartAddr = (BYTE*)emitLastIns - m_debugInfoSize;
+
+    if ((emitCurIGfreeBase <= (BYTE*)lastInsActualStartAddr) && ((BYTE*)lastInsActualStartAddr < emitCurIGfreeEndp))
+    {
+        // The last instruction is in the current buffer. That means the current IG is non-empty.
+        assert(emitCurIGnonEmpty());
+        assert((BYTE*)lastInsActualStartAddr < emitCurIGfreeNext);
+        assert(emitCurIGinsCnt >= 1);
+        assert(emitCurIGsize >= emitLastIns->idCodeSize());
+
+        size_t insSize    = emitCurIGfreeNext - (BYTE*)lastInsActualStartAddr;
+        emitCurIGfreeNext = (BYTE*)lastInsActualStartAddr;
+        emitCurIGinsCnt -= 1;
+        emitCurIGsize -= emitLastIns->idCodeSize();
+
+        // We're going to overwrite the memory; zero it.
+        memset(emitCurIGfreeNext, 0, insSize);
+    }
+    else
+    {
+        // The last instruction has already been saved. It must be the last instruction in the group.
+        assert((BYTE*)emitLastInsIG->igData + emitLastInsIG->igDataSize ==
+               (BYTE*)emitLastIns + emitSizeOfInsDsc(emitLastIns));
+        assert(emitLastInsIG->igInsCnt >= 1);
+        assert(emitLastInsIG->igSize >= emitLastIns->idCodeSize());
+
+        emitLastInsIG->igInsCnt -= 1;
+        emitLastInsIG->igSize -= (unsigned short)emitLastIns->idCodeSize();
+
+        // We don't overwrite this memory; it's simply ignored. We could zero it to be sure nobody uses it,
+        // but it's not necessary, and leaving it might be useful for debugging.
+    }
+
+    // We no longer know what the previous instruction is.
+    emitLastIns = nullptr;
```

Comment:
@BruceForstall
We are having a little trouble under some circumstances after an
instruction is removed, as we enter emitSavIG during the writing
of the optimized LDP or STP instruction. (I am not entirely sure **why**
this is yet :) ) There is an assertion within emitSavIG
to ensure that emitLastIns is not a null pointer.
Do you think we need to keep hold of the instruction
before emitLastIns so we can restore back to that?
Also, if the instruction that is to be removed actually **is** the first
instruction in the group, then emitLastIns **will** be a null pointer,
if I understand correctly. Again, I would appreciate your comments.

Reply:
@AndyJGraham I'm recently back from vacation, so I'll look at this soon.

Can you make sure the PR is up-to-date with your changes (i.e., can you push to GitHub any additional changes you've made locally), so I can build and debug your branch?

Also, is there a specific test or SuperPMI replay that fails with these changes, in the way you describe?

Reply:
@BruceForstall I have made some progress with this:
Problem #1 was that setting the emitLastIns to nullptr caused an assertion to be raised under some circumstances. This has been overcome by recording the last but one instruction and using that rather than nullptr.
Problem #2 was that some sequences of ldr instructions overwrote the register against which we were indexing, so the the second instruction ended up having the wrong register value when it indexed against it. A very simple fix to check for register numbers cured that.
Problem #3 is that under some circumstances, there is a search for an instruction number that is no longer within the instruction group following instruction removal. In the example I have, instruction #7 in a group with only 6 instructions, as the optimized instruction has been put into the next instruction group. This is the problem that I am working on at present.

Reply:
Attached is the C# program (generated by Fuzzlyn) that I use to recreate the error.
[Program.zip](https://github.com/dotnet/runtime/files/10115063/Program.zip)